Update Zig page (#1195)
fermyon/developer: content/wasm-languages/zig.md

## Available Implementations

Zig [supports building for WebAssembly](https://ziglang.org/documentation/0.11.0/#WebAssembly) out of the box.

## Usage

> Zig supports [generating code for all targets that LLVM supports](https://ziglang.org/documentation/0.11.0/#Targets). LLVM has a `wasm32-wasi` target, so Zig should be usable to build Fermyon Platform applications.

Reviewer: Are we able to say "is" rather than "should be" usable?

Things we like:

- The Zig compiler is so nice that we now use it to compile our C and Zig.

Things we're not big fans of:

> - When we get Wasm-related errors, they can be really terse. However, since the original publishing of this page (around Zig 0.4.0), updates have been made to the error section of the official Zig documentation e.g. explaining how [Zig's error handling](https://ziglang.org/documentation/0.11.0/#toc-Common-errdefer-Slip-Ups) includes `defer` statements and `errdefer`, which triggers on block-exit errors. For example:

Reviewer: I'm not sure "`errdefer` slip-ups" is really relevant to Wasm errors being terse?

> - When we get Wasm-related errors, they can be really terse. However, since the original publishing of this page (around Zig 0.4.0), updates have been made to the error section of the official Zig documentation e.g. explaining how [Zig's error handling](https://ziglang.org/documentation/0.11.0/#toc-Common-errdefer-Slip-Ups) includes `defer` statements and `errdefer`, which triggers on block-exit errors. For example:

Reviewer: I wouldn't bother discussing the publishing history - talk about the current situation.

```

> fn createFoo(param: i32) !Foo {

Reviewer: I am not getting what this example is trying to show.

    const foo = try tryToAllocateFoo();
    // now we have allocated foo. we need to free it if the function fails.
    // but we want to return it if the function succeeds.
    errdefer deallocateFoo(foo);

    const tmp_buf = allocateTmpBuffer() orelse return error.OutOfMemory;
    // tmp_buf is truly a temporary resource, and we for sure want to clean it up
    // before this block leaves scope
    defer deallocateTmpBuffer(tmp_buf);

    if (param > 1337) return error.InvalidParam;

    // here the errdefer will not run since we're returning success from the function.
    // but the defer will run!
    return foo;
}
```

# Examples

> >> All of our examples follow [a documented pattern using common tools](/wasm-languages/about-examples). In these examples, we initially run our applications using the [wasmtime](https://github.com/bytecodealliance/wasmtime) runtime and then progress and demonstrate how to run Zig applications using the [Spin framework](https://developer.fermyon.com/spin/v2/index).

Reviewer: ```suggestion
>> All of our examples follow [a documented pattern using common tools](/wasm-languages/about-examples). In these examples, we initially run our applications using the [wasmtime](https://github.com/bytecodealliance/wasmtime) runtime and then progress and demonstrate how to run Zig applications using the [Spin framework](/spin/v2/index).
```

Author: Ended up adding `../spin/v2/*` because we are one layer down here in `wasm-languages`.
Thanks for spotting this absolute / relative issue 💯

> You must have [Zig](https://ziglang.org/learn/) installed. If you haven't already done so, please also go ahead and [install Spin](https://developer.fermyon.com/spin/v2/install).

Reviewer: ```suggestion
You must have [Zig](https://ziglang.org/learn/) installed. If you haven't already done so, please also go ahead and [install Spin](/spin/v2/install).
```

## Example 1

Below is a WebAssembly System Interface (WASI) example where Zig uses the standard library to read command line arguments.

Create a new Zig program:

```console
$ mkdir hello-zig
$ cd hello-zig
$ zig init-exe
info: Created build.zig
info: Created src/main.zig
info: Next, try `zig build --help` or `zig build run`
```

Inside the `src/` directory is a file named `main.zig`. Edit it as follows:

```
const std = @import("std");

pub fn main() !void {
    var general_purpose_allocator = std.heap.GeneralPurposeAllocator(.{}){};
    const gpa = general_purpose_allocator.allocator();

>     const args = try std.process.argsAlloc(gpa);

Reviewer: Wow, this is the most terrifying "print the command line args" program I have ever seen

    defer std.process.argsFree(gpa, args);

    for (args, 0..) |arg, i| {
        std.debug.print("{}: {s}\n", .{ i, arg });
    }
}
```

We can now build and run the WASI example:

```console
$ zig build-exe src/main.zig -target wasm32-wasi
```

As mentioned above, we are going to run our application using the [wasmtime](https://github.com/bytecodealliance/wasmtime) runtime:

```console
$ wasmtime src/main.wasm 123 hello
0: main.wasm
1: 123
2: hello
```

## Example 2

> In this example we are going to build a hello-world-type of example and build using Zig. We will then run the application with wasmtime, and then the Zig-built `.wasm` binary will run inside a [Spin](https://developer.fermyon.com/spin/v2/index) application.

Reviewer: ```suggestion
In this example we are going to build a "hello world" WAGI application using Zig. We will then run the Wasm binary with wasmtime to view the output in the console, and then run it again as a [Spin](/spin/v2/index) application served over HTTP.
```

## Example 1

Below is a WebAssembly System Interface (WASI) example where Zig uses the standard library to read command line arguments.

Create a new Zig program:

```bash
$ mkdir hello-zig
$ cd hello-zig
$ zig init-exe
info: Created build.zig
info: Created src/main.zig
info: Next, try `zig build --help` or `zig build run`
```

Inside the `src/` directory is a file named `main.zig`. Edit it as follows:

```
const std = @import("std");

pub fn main() !void {
    var general_purpose_allocator = std.heap.GeneralPurposeAllocator(.{}){};
    const gpa = general_purpose_allocator.allocator();
    const args = try std.process.argsAlloc(gpa);
    defer std.process.argsFree(gpa, args);

    for (args, 0..) |arg, i| {
        std.debug.print("{}: {s}\n", .{ i, arg });
    }
}
```

We can now build and run the WASI example:

```bash
$ zig build-exe src/main.zig -target wasm32-wasi
```

As mentioned above, we are going to run our application using the [wasmtime](https://github.com/bytecodealliance/wasmtime) runtime:

```bash
$ wasmtime src/main.wasm 123 hello
0: main.wasm
1: 123
2: hello
```

## Example 2

> In this example we are going to build a "hello world" WAGI application using Zig. We will then run the Wasm binary with wasmtime to view the output in the console, and then run it again as a [Spin](../spin/v2/index) application served over HTTP.

Reviewer: Is there a reason to use a `..` path rather than a `/` path?  Our experience has been that `..` paths are more prone to getting messed up when we reorganise files (whereas `/` paths usually at least hit a redirect even if stuff has moved).

Author: It is mainly that Visual Studio Code responds to the `../` and provides a path that I can tab into (which makes it feel real[er])

}
```

Now compile and run the program using wasmtime:

```console
$ zig build-exe -O ReleaseSmall -target wasm32-wasi src/main.zig
$ wasmtime main.wasm    
content-type: text/plain
Hello, World!
```

Now you should have a `main.wasm` file:

```console
$ tree .
.
├── build.zig
├── main.wasm
├── main.wasm.o
├── src
│   └── main.zig

> └── zig-cache

Reviewer: Not sure showing the contents of this is helpful - to me it felt scary rather than informative.

    ├── h
    ├── o
    │   └── c8d45f36408768ee488d7b453a8723a3
    │       └── builtin.zig
    └── z
        └── b78f88b24cc6c1a13507ecfccf9702c3
```

### Using Spin

Create a new `spin.toml` file and edit it as follows:

```toml
spin_manifest_version = 2

[application]
name = "spin-hello-zig"
version = "0.1.0"
authors = ["Fermyon Engineering <engineering@fermyon.com>"]
description = "spin-hello-zig"

[[trigger.http]]
route = "/..."
executor = { type = "wagi" } # Note: We are running this using the Wagi spec
component = "spin-hello-zig"

[component.spin-hello-zig]
source = "main.wasm"
```

Once we have the Spin manifest file (`spin.toml`), we can run `spin up`:

```console
spin up     
Logging component stdio to ".spin/logs/"

Serving http://127.0.0.1:3000
Available Routes:
  spin-hello-zig: http://127.0.0.1:3000 (wildcard)
```

Then, use `curl` (in a new terminal) or a web browser to send a request:

> ```curl

Reviewer: This one has gone whackadoo in the preview too...!
